diff --git a/setup.py b/setup.py
index c6ec32bf..cd2fdba6 100644
--- a/setup.py
+++ b/setup.py
@@ -30,7 +30,7 @@ def _get_version():
         with open(os.path.join(ROOT_DIR, "version.txt"), "r") as f:
             version = f.readline().strip()
         if sha is not None and "OFFICIAL_RELEASE" not in os.environ:
-            version += "+" + sha[:7]
+            pass

     if sha is None:
         sha = "Unknown"
@@ -86,13 +86,13 @@ def main(argv: List[str]) -> None:
         name="torchrec",
         version=version,
         author="TorchRec Team",
-        author_email="packages@pytorch.org",
-        maintainer="TroyGarden",
-        maintainer_email="hhy@meta.com",
+        author_email="",
+        maintainer="",
+        maintainer_email="",
         description="TorchRec: Pytorch library for recommendation systems",
         long_description=readme,
         long_description_content_type="text/markdown",
-        url="https://github.com/meta-pytorch/torchrec",
+        url="",
         license="BSD-3",
         keywords=[
             "pytorch",
diff --git a/torchrec/__init__.py b/torchrec/__init__.py
index 29258f6f..ab36dc52 100644
--- a/torchrec/__init__.py
+++ b/torchrec/__init__.py
@@ -8,4 +8,16 @@
 # pyre-strict
-
-import torchrec.distributed  # noqa
+
+import enum
+from fbgemm_gpu import split_table_batched_embeddings_ops_training
+
+
+class ComputeDevice(enum.IntEnum):
+    CPU = 0
+    CUDA = 1
+    MTIA = 2
+    NPU = 3
+
+split_table_batched_embeddings_ops_training.ComputeDevice = ComputeDevice
+
+import torchrec.distributed  # noqa
 import torchrec.quant  # noqa
diff --git a/torchrec/distributed/batched_embedding_kernel.py b/torchrec/distributed/batched_embedding_kernel.py
index e0316ef2..d7e4fcf1 100644
--- a/torchrec/distributed/batched_embedding_kernel.py
+++ b/torchrec/distributed/batched_embedding_kernel.py
@@ -29,6 +29,9 @@ from typing import (
     TypeVar,
     Union,
 )
+import uuid
+from itertools import accumulate
+from math import log2

 import torch
 import torch.distributed as dist
@@ -54,6 +57,7 @@ from fbgemm_gpu.tbe.ssd import ASSOC, SSDTableBatchedEmbeddingBags
 from fbgemm_gpu.tbe.ssd.utils.partially_materialized_tensor import (
     PartiallyMaterializedTensor,
 )
+from fbgemm_gpu.config import FeatureGate, FeatureGateName
 from torch import nn
 from torch.autograd.profiler import record_function
 from torch.distributed._tensor import DTensor, Replicate, Shard as DTensorShard
@@ -161,6 +165,136 @@ class ReduceScatterResizeAwaitable(LazyAwaitable[torch.Tensor]):
         return self._shard_buf


+def custom_init(
+        self,
+        embedding_specs: List[Tuple[int, int]],  # tuple of (rows, dims)
+        feature_table_map: Optional[List[int]] = None,  # [T]
+        weights_precision: SparseType = SparseType.FP32,
+        pooling_mode: PoolingMode = PoolingMode.SUM,
+        use_cpu: bool = False,
+        output_dtype: SparseType = SparseType.FP32,
+        use_mtia: bool = False,
+) -> None:  # noqa C901  # tuple of (rows, dims,)
+    super(DenseTableBatchedEmbeddingBagsCodegen, self).__init__()
+    self.uuid = str(uuid.uuid4())
+
+    self.log(
+        f"Feature Gates: {[(feature.name, feature.is_enabled()) for feature in FeatureGateName]}"
+    )
+
+    self.pooling_mode = pooling_mode
+    self.weights_precision = weights_precision
+    self.output_dtype: int = output_dtype.as_int()
+    table_embedding_dtype = weights_precision.as_dtype()
+
+    self.use_cpu: bool = use_cpu
+    self.use_mtia: bool = use_mtia
+
+    assert not (use_cpu and use_mtia), "Cannot use CPU and MTIA at the same time"
+
+    if self.use_cpu or self.pooling_mode == PoolingMode.NONE:
+        assert output_dtype in [
+            SparseType.FP32,
+            SparseType.FP16,
+            SparseType.BF16,
+        ], "Fused pooled embedding quantization only supported for cuda."
+
+    # pyre-fixme[8]: Attribute has type `device`; used as `Union[int, device]`.
+    self.current_device: torch.device = (
+        torch.device("cpu")
+        if self.use_cpu
+        else (
+            torch.device(f"mtia:{torch.mtia.current_device()}")
+            if self.use_mtia
+            # support npu
+            else (torch.npu.current_device() if torch.npu.is_available() else torch.cuda.current_device())
+        )
+    )
+
+    self.embedding_specs = embedding_specs
+    (rows, dims) = zip(*embedding_specs)
+    T_ = len(self.embedding_specs)
+    assert T_ > 0
+
+    feature_table_map = (
+        feature_table_map if feature_table_map is not None else list(range(T_))
+    )
+    T = len(feature_table_map)
+    assert T_ <= T
+
+    feature_dims = [dims[t] for t in feature_table_map]
+    D_offsets = [0] + list(accumulate(feature_dims))
+    self.total_D = D_offsets[-1]
+    self.max_D = max(dims)
+    self.register_buffer(
+        "D_offsets",
+        torch.tensor(D_offsets, device=self.current_device, dtype=torch.int32),
+    )
+    assert self.D_offsets.numel() == T + 1
+
+    # Required for VBE
+    self.register_buffer(
+        "feature_dims",
+        torch.tensor(feature_dims, device="cpu", dtype=torch.int64),
+    )
+
+    hash_size_cumsum = [0] + list(accumulate(rows))
+    if hash_size_cumsum[-1] == 0:
+        self.total_hash_size_bits: int = 0
+    else:
+        self.total_hash_size_bits: int = int(log2(float(hash_size_cumsum[-1])) + 1)
+    # The last element is to easily access # of rows of each table by
+    # hash_size_cumsum[t + 1] - hash_size_cumsum[t]
+    hash_size_cumsum = [hash_size_cumsum[t] for t in feature_table_map] + [
+        hash_size_cumsum[-1]
+    ]
+    self.register_buffer(
+        "hash_size_cumsum",
+        torch.tensor(
+            hash_size_cumsum, device=self.current_device, dtype=torch.int64
+        ),
+    )
+    weights_offsets = [0] + list(
+        accumulate([row * dim for (row, dim) in embedding_specs])
+    )
+    self.weights = nn.Parameter(
+        torch.randn(
+            weights_offsets[-1],
+            device=self.current_device,
+            dtype=table_embedding_dtype,
+        )
+    )
+    for feature in range(T):
+        t = feature_table_map[feature]
+        row, dim = embedding_specs[t]
+        if (
+                self.weights[weights_offsets[t] : weights_offsets[t + 1]].numel()
+                != row * dim
+        ):
+            self.log(
+                f"row {row} dim {dim} feature {feature} t {t} {self.weights[weights_offsets[t] : weights_offsets[t + 1]].numel()}"
+            )
+        assert (
+                self.weights[weights_offsets[t] : weights_offsets[t + 1]].numel()
+                == row * dim
+        )
+        assert self.hash_size_cumsum[feature] == sum(
+            row for (row, _) in embedding_specs[:t]
+        )
+
+    self.weights_physical_offsets: List[int] = weights_offsets
+    weights_offsets = [weights_offsets[t] for t in feature_table_map]
+    self.register_buffer(
+        "weights_offsets",
+        torch.tensor(
+            weights_offsets, device=self.current_device, dtype=torch.int64
+        ),
+    )
+
+
+DenseTableBatchedEmbeddingBagsCodegen.__init__ = custom_init
+
+
 def _populate_res_params(config: GroupedEmbeddingConfig) -> Tuple[bool, RESParams]:
     # populate res_params, which is used for raw embedding streaming
     # here only populates the params available in fused_params and TBE configs
@@ -2433,6 +2567,11 @@ class BatchedFusedEmbedding(BaseBatchedEmbedding[torch.Tensor], FusedOptimizerMo
                 # Set EmbeddingLocation.HOST to make embedding op in FBGEMM choose CPU path.
                 # But the tensor will still be created on MTIA with device type "mtia".
                 managed.append(EmbeddingLocation.HOST)
+            elif device is not None and device.type == "npu":
+                compute_devices.append(ComputeDevice.NPU)
+                managed.append(
+                    compute_kernel_to_embedding_location(table.compute_kernel)
+                )
             else:
                 compute_devices.append(ComputeDevice.CPU)
                 managed.append(EmbeddingLocation.HOST)
@@ -2751,7 +2890,7 @@ class BatchedDenseEmbedding(BaseBatchedEmbedding[torch.Tensor]):
         use_cpu: bool = (
             device is None
             or device.type == "cpu"
-            or (not (torch.cuda.is_available() or torch.mtia.is_available()))
+            or (not (torch.cuda.is_available() or torch.mtia.is_available() or torch.npu.is_available()))
         )
         self._emb_module: DenseTableBatchedEmbeddingBagsCodegen = (
             DenseTableBatchedEmbeddingBagsCodegen(
@@ -3544,6 +3683,11 @@ class BatchedFusedEmbeddingBag(
                 # Set EmbeddingLocation.HOST to make embedding op in FBGEMM choose CPU path.
                 # But the tensor will still be created on MTIA with device type "mtia".
                 managed.append(EmbeddingLocation.HOST)
+            elif device is not None and device.type == "npu":
+                compute_devices.append(ComputeDevice.NPU)
+                managed.append(
+                    compute_kernel_to_embedding_location(table.compute_kernel)
+                )
             else:
                 compute_devices.append(ComputeDevice.CPU)
                 managed.append(EmbeddingLocation.HOST)
@@ -3865,7 +4009,7 @@ class BatchedDenseEmbeddingBag(BaseBatchedEmbeddingBag[torch.Tensor]):
         use_cpu: bool = (
             device is None
             or device.type == "cpu"
-            or (not (torch.cuda.is_available() or torch.mtia.is_available()))
+            or (not (torch.cuda.is_available() or torch.mtia.is_available() or torch.npu.is_available()))
         )
         self._emb_module: DenseTableBatchedEmbeddingBagsCodegen = (
             DenseTableBatchedEmbeddingBagsCodegen(
diff --git a/torchrec/distributed/benchmark/benchmark_comms.py b/torchrec/distributed/benchmark/benchmark_comms.py
index 483463f7..1a8574cd 100644
--- a/torchrec/distributed/benchmark/benchmark_comms.py
+++ b/torchrec/distributed/benchmark/benchmark_comms.py
@@ -574,7 +574,7 @@ def a2a_single_runner(rank: int, world_size: int, arg: AllToAllSingleRunConfig)
     with MultiProcessContext(
         rank=rank,
         world_size=world_size,
-        backend="nccl",
+        backend="hccl" if torch.npu.is_available() else "nccl",
         use_deterministic_algorithms=False,
     ) as ctx:
         match arg.name.lower():
diff --git a/torchrec/distributed/benchmark/benchmark_split_table_batched_embeddings.py b/torchrec/distributed/benchmark/benchmark_split_table_batched_embeddings.py
index 07e0de20..5058679b 100644
--- a/torchrec/distributed/benchmark/benchmark_split_table_batched_embeddings.py
+++ b/torchrec/distributed/benchmark/benchmark_split_table_batched_embeddings.py
@@ -56,17 +56,19 @@ def op_bench(
     batch_size: int,
     bag_size: int,
 ) -> None:
+    if torch.npu.is_available():
+        compute_device = ComputeDevice.NPU
+    elif torch.cuda.is_available():
+        compute_device = ComputeDevice.CUDA
+    else:
+        compute_device = ComputeDevice.CPU
     emb = SplitTableBatchedEmbeddingBagsCodegen(
         [
             (
                 num_embeddings,
                 embedding_dim,
                 EmbeddingLocation.DEVICE,
-                (
-                    ComputeDevice.CUDA
-                    if torch.cuda.is_available()
-                    else ComputeDevice.CPU
-                ),
+                compute_device,
             )
         ]
         * num_tables,
@@ -96,6 +98,7 @@ def op_bench(
             feature_names=["feature_0"],
         )
     ]
+    device_type = "npu" if torch.npu.is_available() else "cuda"
     inputs = ModelInput.generate(
         tables=tables,
         weighted_tables=[],
@@ -103,7 +106,7 @@ def op_bench(
         world_size=1,
         num_float_features=0,
         pooling_avg=10,
-        device=torch.device("cuda"),
+        device=torch.device(device_type),
     )[0].idlist_features

     result = benchmark_func(
diff --git a/torchrec/distributed/benchmark/benchmark_train.py b/torchrec/distributed/benchmark/benchmark_train.py
index bff193d2..e243dfc1 100644
--- a/torchrec/distributed/benchmark/benchmark_train.py
+++ b/torchrec/distributed/benchmark/benchmark_train.py
@@ -84,14 +84,24 @@ def benchmark_ebc(
     # to CPU because this module will be sent over subprocesses via multiprocessing,
     # and we don't want to create an extra CUDA context on GPU0 for each subprocess.
     # we also need to release the memory in the parent process (empty_cache)
-    module = EmbeddingBagCollection(
-        # pyre-ignore [6]
-        tables=table_configs,
-        is_weighted=False,
-        device=torch.device("cuda"),
-    ).cpu()
-
-    torch.cuda.empty_cache()
+    if torch.npu.is_available():
+        module = EmbeddingBagCollection(
+            # pyre-ignore [6]
+            tables=table_configs,
+            is_weighted=False,
+            device=torch.device("npu"),
+        ).cpu()
+
+        torch.npu.empty_cache()
+    else:
+        module = EmbeddingBagCollection(
+            # pyre-ignore [6]
+            tables=table_configs,
+            is_weighted=False,
+            device=torch.device("cuda"),
+        ).cpu()
+
+        torch.cuda.empty_cache()

     IGNORE_ARGNAME = ["output_dir", "embedding_config_json", "max_num_embeddings"]
     optimizer = torch.optim.SGD(module.parameters(), lr=0.02)
diff --git a/torchrec/distributed/benchmark/benchmark_train_pipeline.py b/torchrec/distributed/benchmark/benchmark_train_pipeline.py
index ab0ab3c3..a1c0acf3 100644
--- a/torchrec/distributed/benchmark/benchmark_train_pipeline.py
+++ b/torchrec/distributed/benchmark/benchmark_train_pipeline.py
@@ -132,14 +132,14 @@ def runner(
 ) -> BenchmarkResult:
     # Ensure GPUs are available and we have enough of them
     assert (
-        torch.cuda.is_available() and torch.cuda.device_count() >= world_size
-    ), "CUDA not available or insufficient GPUs for the requested world_size"
+        torch.npu.is_available() and torch.npu.device_count() >= world_size
+    ), "NPU not available or insufficient GPUs for the requested world_size"

     run_option.set_log_level()
     with MultiProcessContext(
         rank=rank,
         world_size=world_size,
-        backend="nccl",
+        backend="hccl" if torch.npu.is_available() else "nccl",
         use_deterministic_algorithms=False,
     ) as ctx:
         unsharded_model = model_config.generate_model(
diff --git a/torchrec/distributed/comm_ops.py b/torchrec/distributed/comm_ops.py
index f9fcbaeb..688be58d 100644
--- a/torchrec/distributed/comm_ops.py
+++ b/torchrec/distributed/comm_ops.py
@@ -157,7 +157,7 @@ class Request(Awaitable[W]):
         """

         ret = self.wait_function.apply(self.pg, self, self.dummy_tensor)
-        if isinstance(ret, torch.Tensor) and ret.device.type == "cuda":
+        if isinstance(ret, torch.Tensor) and (ret.device.type == "cuda" or ret.device.type == "npu"):
             ret.record_stream(torch.get_device_module(ret.device).current_stream())
         self.req = None
         self.tensor = None
diff --git a/torchrec/distributed/dist_data.py b/torchrec/distributed/dist_data.py
index 99ed9e04..0006f43f 100644
--- a/torchrec/distributed/dist_data.py
+++ b/torchrec/distributed/dist_data.py
@@ -794,7 +794,7 @@ class KJTOneToAll(nn.Module):
             # BUG: device will default to cuda if cpu specified
             self._device_type: str = (
                 device.type
-                if device is not None and device.type in {"meta", "cuda", "mtia"}
+                if device is not None and device.type in {"meta", "cuda", "mtia", "npu"}
                 else "cuda"
             )
         assert self._world_size == len(splits)
diff --git a/torchrec/distributed/embedding.py b/torchrec/distributed/embedding.py
index a6e7976b..973f138e 100644
--- a/torchrec/distributed/embedding.py
+++ b/torchrec/distributed/embedding.py
@@ -294,7 +294,7 @@ def _pin_and_move(tensor: torch.Tensor, device: torch.device) -> torch.Tensor:
     """
     return (
         tensor.pin_memory().to(device=device, non_blocking=True)
-        if device.type == "cuda" and tensor.device.type == "cpu"
+        if (device.type == "cuda" or device.type == "npu") and tensor.device.type == "cpu"
         else tensor.to(device=device, non_blocking=True)
     )

@@ -564,7 +564,7 @@ class ShardedEmbeddingCollection(
                     module=lookup,
                     device_ids=(
                         [self._device]
-                        if self._device is not None and self._device.type == "cuda"
+                        if self._device is not None and (self._device.type == "cuda" or self._device.type == "npu")
                         else None
                     ),
                     process_group=env.process_group,
diff --git a/torchrec/distributed/embedding_lookup.py b/torchrec/distributed/embedding_lookup.py
index e628c897..9b8194d7 100644
--- a/torchrec/distributed/embedding_lookup.py
+++ b/torchrec/distributed/embedding_lookup.py
@@ -1498,7 +1498,9 @@ class InferGroupedPooledEmbeddingsLookup(
             )  # TODO: replace hardcoded cpu with DEFAULT_DEVICE_TYPE in torchrec.distributed.types when torch package issue resolved
         else:
             device_type = (
-                "meta" if device is not None and device.type == "meta" else "cuda"
+                "meta"
+                if device is not None and device.type == "meta"
+                else ("npu" if torch.npu.is_available() else "cuda")
             )

         self._is_empty_rank: List[bool] = []
@@ -1580,7 +1582,9 @@ class InferGroupedEmbeddingsLookup(
             )  # TODO: replace hardcoded cpu with DEFAULT_DEVICE_TYPE in torchrec.distributed.types when torch package issue resolved
         else:
             device_type = (
-                "meta" if device is not None and device.type == "meta" else "cuda"
+                "meta"
+                if device is not None and device.type == "meta"
+                else ("npu" if torch.npu.is_available() else "cuda")
             )
         for rank in range(world_size):
             grouped_configs_per_rank_elem = grouped_configs_per_rank[rank]
diff --git a/torchrec/distributed/embedding_types.py b/torchrec/distributed/embedding_types.py
index aa50640f..f260d54d 100644
--- a/torchrec/distributed/embedding_types.py
+++ b/torchrec/distributed/embedding_types.py
@@ -537,7 +537,7 @@ class BaseEmbeddingSharder(ModuleSharder[M]):
             ShardingType.COLUMN_WISE.value,
             ShardingType.TABLE_COLUMN_WISE.value,
         ]
-        if compute_device_type in {"cuda"}:
+        if compute_device_type in {"cuda", "npu"}:
             types += [
                 ShardingType.ROW_WISE.value,
                 ShardingType.TABLE_ROW_WISE.value,
@@ -557,7 +557,7 @@ class BaseEmbeddingSharder(ModuleSharder[M]):
             ret += [
                 EmbeddingComputeKernel.FUSED.value,
             ]
-            if compute_device_type in {"cuda"}:
+            if compute_device_type in {"cuda", "npu"}:
                 ret += [
                     EmbeddingComputeKernel.FUSED_UVM.value,
                     EmbeddingComputeKernel.FUSED_UVM_CACHING.value,
@@ -588,14 +588,15 @@ class BaseEmbeddingSharder(ModuleSharder[M]):
             EmbeddingComputeKernel.FUSED_UVM.value,
             EmbeddingComputeKernel.FUSED_UVM_CACHING.value,
         }:
-            assert compute_device_type in {"cuda"}
+            assert compute_device_type in {"cuda", "npu"}
             return {ParameterStorage.DDR.value: tensor_bytes}
         else:
-            assert compute_device_type in {"cuda", "cpu", "mtia"}
+            assert compute_device_type in {"cuda", "cpu", "mtia", "npu"}
             storage_map = {
                 "cuda": ParameterStorage.HBM,
                 "cpu": ParameterStorage.DDR,
                 "mtia": ParameterStorage.HBM,
+                "npu": ParameterStorage.HBM
             }
             return {
                 storage_map[compute_device_type].value: get_tensor_size_bytes(tensor)
@@ -668,7 +669,7 @@ class BaseQuantEmbeddingSharder(ModuleSharder[M]):
         ret = [
             EmbeddingComputeKernel.QUANT.value,
         ]
-        if compute_device_type in {"cuda"}:
+        if compute_device_type in {"cuda", "npu"}:
             ret += [
                 EmbeddingComputeKernel.QUANT_UVM.value,
                 EmbeddingComputeKernel.QUANT_UVM_CACHING.value,
@@ -691,14 +692,15 @@ class BaseQuantEmbeddingSharder(ModuleSharder[M]):
             EmbeddingComputeKernel.QUANT_UVM.value,
             EmbeddingComputeKernel.QUANT_UVM_CACHING.value,
         }:
-            assert compute_device_type in {"cuda"}
+            assert compute_device_type in {"cuda", "npu"}
             return {ParameterStorage.DDR.value: tensor_bytes}
         else:
-            assert compute_device_type in {"cuda", "cpu", "mtia"}
+            assert compute_device_type in {"cuda", "cpu", "mtia", "npu"}
             storage_map = {
                 "cuda": ParameterStorage.HBM,
                 "cpu": ParameterStorage.DDR,
                 # TODO: Update it later. Setting for MTIA is same as CPU's for now.
                 "mtia": ParameterStorage.DDR,
+                "npu": ParameterStorage.HBM
             }
             return {storage_map[compute_device_type].value: tensor_bytes}
diff --git a/torchrec/distributed/embeddingbag.py b/torchrec/distributed/embeddingbag.py
index fca8a9c5..143c63f7 100644
--- a/torchrec/distributed/embeddingbag.py
+++ b/torchrec/distributed/embeddingbag.py
@@ -638,7 +638,7 @@ class ShardedEmbeddingBagCollection(
                     device_ids=(
                         [self._device]
                         if self._device is not None
-                        and (self._device.type in {"cuda", "mtia"})
+                        and (self._device.type in {"cuda", "mtia", "npu"})
                         else None
                     ),
                     process_group=env.process_group,
diff --git a/torchrec/distributed/fbgemm_qcomm_codec.py b/torchrec/distributed/fbgemm_qcomm_codec.py
index 1a93b6ff..e98fd6c6 100644
--- a/torchrec/distributed/fbgemm_qcomm_codec.py
+++ b/torchrec/distributed/fbgemm_qcomm_codec.py
@@ -212,7 +212,7 @@ def get_qcomm_codecs_registry(
         return None

     if device is None:
-        device = torch.device("cuda")
+        device = torch.device("npu" if torch.npu.is_available() else "cuda")

     qcomm_codecs_registry = {}
     if comm_ops is None:
diff --git a/torchrec/distributed/fp_embeddingbag.py b/torchrec/distributed/fp_embeddingbag.py
index 40c13db8..1c9fb6cb 100644
--- a/torchrec/distributed/fp_embeddingbag.py
+++ b/torchrec/distributed/fp_embeddingbag.py
@@ -242,7 +242,7 @@ class FeatureProcessedEmbeddingBagCollectionSharder(
     ) -> ShardedFeatureProcessedEmbeddingBagCollection:

         if device is None:
-            device = torch.device("cuda")
+            device = torch.device("npu" if torch.npu.is_available() else "cuda")

         return ShardedFeatureProcessedEmbeddingBagCollection(
             module,
diff --git a/torchrec/distributed/fused_embedding.py b/torchrec/distributed/fused_embedding.py
index 15b731ad..42b4f6b4 100644
--- a/torchrec/distributed/fused_embedding.py
+++ b/torchrec/distributed/fused_embedding.py
@@ -134,7 +134,7 @@ class FusedEmbeddingCollectionSharder(BaseEmbeddingSharder[FusedEmbeddingCollect
             ret += [
                 EmbeddingComputeKernel.FUSED.value,
             ]
-            if compute_device_type in {"cuda"}:
+            if compute_device_type in {"cuda", "npu"}:
                 ret += [
                     EmbeddingComputeKernel.FUSED_UVM.value,
                     EmbeddingComputeKernel.FUSED_UVM_CACHING.value,
diff --git a/torchrec/distributed/fused_embeddingbag.py b/torchrec/distributed/fused_embeddingbag.py
index 231e3716..e6cd025c 100644
--- a/torchrec/distributed/fused_embeddingbag.py
+++ b/torchrec/distributed/fused_embeddingbag.py
@@ -127,7 +127,7 @@ class FusedEmbeddingBagCollectionSharder(
             ShardingType.COLUMN_WISE.value,
             ShardingType.TABLE_COLUMN_WISE.value,
         ]
-        if compute_device_type in {"cuda"}:
+        if compute_device_type in {"cuda", "npu"}:
             types += [
                 ShardingType.ROW_WISE.value,
                 ShardingType.TABLE_ROW_WISE.value,
@@ -143,7 +143,7 @@ class FusedEmbeddingBagCollectionSharder(
             ret += [
                 EmbeddingComputeKernel.FUSED.value,
             ]
-            if compute_device_type in {"cuda"}:
+            if compute_device_type in {"cuda", "npu"}:
                 ret += [
                     EmbeddingComputeKernel.FUSED_UVM.value,
                     EmbeddingComputeKernel.FUSED_UVM_CACHING.value,
diff --git a/torchrec/distributed/itep_embeddingbag.py b/torchrec/distributed/itep_embeddingbag.py
index e3db9553..9b8ac5a2 100644
--- a/torchrec/distributed/itep_embeddingbag.py
+++ b/torchrec/distributed/itep_embeddingbag.py
@@ -309,7 +309,7 @@ class ITEPEmbeddingBagCollectionSharder(

         # Enforce GPU for ITEPEmbeddingBagCollection
         if device is None:
-            device = torch.device("cuda")
+            device = torch.device("npu" if torch.npu.is_available() else "cuda")

         return ShardedITEPEmbeddingBagCollection(
             module,
@@ -568,7 +568,7 @@ class ITEPEmbeddingCollectionSharder(BaseEmbeddingSharder[ITEPEmbeddingCollectio

         # Enforce GPU for ITEPEmbeddingBagCollection
         if device is None:
-            device = torch.device("cuda")
+            device = torch.device("npu" if torch.npu.is_available() else "cuda")

         return ShardedITEPEmbeddingCollection(
             module,
diff --git a/torchrec/distributed/keyed_jagged_tensor_pool.py b/torchrec/distributed/keyed_jagged_tensor_pool.py
index 4ab917ea..672f45e5 100644
--- a/torchrec/distributed/keyed_jagged_tensor_pool.py
+++ b/torchrec/distributed/keyed_jagged_tensor_pool.py
@@ -210,7 +210,7 @@ class ShardedKeyedJaggedTensorPool(
         self._pool_size = pool_size
         self._values_dtype = values_dtype
         self._sharding_env = sharding_env
-        self._device: torch.device = device or torch.device("cuda")
+        self._device: torch.device = device or torch.device("npu" if torch.npu.is_available() else "cuda")
         self._is_weighted = is_weighted
         self._sharding_plan = sharding_plan
         self._feature_max_lengths = feature_max_lengths
@@ -510,7 +510,7 @@ class ShardedInferenceKeyedJaggedTensorPool(
         self._values_dtype = values_dtype
         self._sharding_env = sharding_env
         self._world_size = self._sharding_env.world_size
-        self._device = device or torch.device("cuda")
+        self._device = device or torch.device("npu" if torch.npu.is_available() else "cuda")
         self._sharding_plan = sharding_plan

         self._is_weighted = is_weighted
@@ -540,7 +540,7 @@ class ShardedInferenceKeyedJaggedTensorPool(
             shard_device = (
                 torch.device("cpu")
                 if device == torch.device("cpu")
-                else torch.device("cuda", rank)
+                else torch.device("npu" if torch.npu.is_available() else "cuda", rank)
             )
             self._local_kjt_pool_shards.append(
                 TensorJaggedIndexSelectLookup(
diff --git a/torchrec/distributed/mc_embedding.py b/torchrec/distributed/mc_embedding.py
index af0315d2..2794afd0 100644
--- a/torchrec/distributed/mc_embedding.py
+++ b/torchrec/distributed/mc_embedding.py
@@ -127,7 +127,7 @@ class ManagedCollisionEmbeddingCollectionSharder(
     ) -> ShardedManagedCollisionEmbeddingCollection:

         if device is None:
-            device = torch.device("cuda")
+            device = torch.device("npu" if torch.npu.is_available() else "cuda")

         return ShardedManagedCollisionEmbeddingCollection(
             module,
diff --git a/torchrec/distributed/mc_embeddingbag.py b/torchrec/distributed/mc_embeddingbag.py
index 35b5293f..1e4f261f 100644
--- a/torchrec/distributed/mc_embeddingbag.py
+++ b/torchrec/distributed/mc_embeddingbag.py
@@ -155,7 +155,7 @@ class ManagedCollisionEmbeddingBagCollectionSharder(
     ) -> ShardedManagedCollisionEmbeddingBagCollection:

         if device is None:
-            device = torch.device("cuda")
+            device = torch.device("npu" if torch.npu.is_available() else "cuda")

         return ShardedManagedCollisionEmbeddingBagCollection(
             module,
diff --git a/torchrec/distributed/model_parallel.py b/torchrec/distributed/model_parallel.py
index ec6e6a8f..8bcb44b9 100644
--- a/torchrec/distributed/model_parallel.py
+++ b/torchrec/distributed/model_parallel.py
@@ -972,8 +972,8 @@ class DMPCollection(DistributedModelParallel):
         rs_awaitable_hook_module: Optional[str] = None,
     ) -> None:
         assert (
-            device.type == "cuda" or device.type == "mtia"
-        ), "DMPCollection only supports CUDA or MTIA"
+            device.type == "npu" or device.type == "cuda" or device.type == "mtia"
+        ), "DMPCollection only supports CUDA or MTIA or NPU"
         self._device = device
         self._pg: dist.ProcessGroup = global_pg
         self._global_rank: int = dist.get_rank(global_pg)
diff --git a/torchrec/distributed/planner/constants.py b/torchrec/distributed/planner/constants.py
index 56c7dc26..9905da0e 100644
--- a/torchrec/distributed/planner/constants.py
+++ b/torchrec/distributed/planner/constants.py
@@ -22,7 +22,8 @@ NUM_POOLINGS: float = 1.0

 BIGINT_DTYPE: int = 8

-HBM_CAP: int = 32 * 1024 * 1024 * 1024  # 32 GB
+# 950PR: 128GB, 950DT: 144GB
+HBM_CAP: int = 128 * 1024 * 1024 * 1024  # 128 GB
 DDR_CAP: int = 128 * 1024 * 1024 * 1024  # 128 GB
 DDR_MEM_BW: float = 51 * 1024 * 1024 * 1024 / 1000  # bytes/ms
 HBM_MEM_BW: float = 897 * 1024 * 1024 * 1024 / 1000  # bytes/ms
@@ -98,6 +99,23 @@ def kernel_bw_lookup(
         ("cuda", EmbeddingComputeKernel.KEY_VALUE.value): hbm_to_ddr_mem_bw,
         ("cuda", EmbeddingComputeKernel.SSD_VIRTUAL_TABLE.value): hbm_to_ddr_mem_bw,
         ("cuda", EmbeddingComputeKernel.DRAM_VIRTUAL_TABLE.value): hbm_to_ddr_mem_bw,
+        # NPU
+        ("npu", EmbeddingComputeKernel.DENSE.value): 0.5 * hbm_mem_bw,
+        ("npu", EmbeddingComputeKernel.FUSED.value): 1 * hbm_mem_bw,
+        ("npu", EmbeddingComputeKernel.FUSED_UVM.value): hbm_to_ddr_mem_bw / 10,
+        ("npu", EmbeddingComputeKernel.FUSED_UVM_CACHING.value): (
+            caching_ratio * hbm_mem_bw + (1 - caching_ratio) * hbm_to_ddr_mem_bw
+        )
+        / 10,
+        ("npu", EmbeddingComputeKernel.QUANT.value): 1 * hbm_mem_bw,
+        ("npu", EmbeddingComputeKernel.QUANT_UVM.value): hbm_to_ddr_mem_bw / 10,
+        ("npu", EmbeddingComputeKernel.QUANT_UVM_CACHING.value): (
+            caching_ratio * hbm_mem_bw + (1 - caching_ratio) * hbm_to_ddr_mem_bw
+        )
+        / 10,
+        ("npu", EmbeddingComputeKernel.KEY_VALUE.value): hbm_to_ddr_mem_bw,
+        ("npu", EmbeddingComputeKernel.SSD_VIRTUAL_TABLE.value): hbm_to_ddr_mem_bw,
+        ("npu", EmbeddingComputeKernel.DRAM_VIRTUAL_TABLE.value): hbm_to_ddr_mem_bw,
     }

     if (
@@ -106,5 +124,11 @@ def kernel_bw_lookup(
         and compute_kernel == EmbeddingComputeKernel.FUSED_UVM_CACHING.value
     ):
         return lookup.get(("cuda", EmbeddingComputeKernel.FUSED.value))
+    if (
+        prefetch_pipeline
+        and compute_device == "npu"
+        and compute_kernel == EmbeddingComputeKernel.FUSED_UVM_CACHING.value
+    ):
+        return lookup.get(("npu", EmbeddingComputeKernel.FUSED.value))

     return lookup.get((compute_device, compute_kernel))
diff --git a/torchrec/distributed/planner/enumerators.py b/torchrec/distributed/planner/enumerators.py
index b24fe2cf..d5b197fa 100644
--- a/torchrec/distributed/planner/enumerators.py
+++ b/torchrec/distributed/planner/enumerators.py
@@ -96,7 +96,7 @@ class EmbeddingEnumerator(Enumerator):
             use_exact_enumerate_order if use_exact_enumerate_order else False
         )
         memory_type = (
-            "hbm_cap" if topology.compute_device in {"cuda", "mtia"} else "ddr_cap"
+            "hbm_cap" if topology.compute_device in {"cuda", "mtia", "npu"} else "ddr_cap"
         )
         self._device_memory_sizes: Optional[
             List[int]
diff --git a/torchrec/distributed/planner/shard_estimators.py b/torchrec/distributed/planner/shard_estimators.py
index 53bb9199..2ee8557f 100644
--- a/torchrec/distributed/planner/shard_estimators.py
+++ b/torchrec/distributed/planner/shard_estimators.py
@@ -1349,7 +1349,7 @@ def calculate_shard_storages(
                 count_ephemeral_storage_cost=count_ephemeral_storage_cost,
                 is_inference=is_inference,
             )
-            if compute_device in {"cuda", "mtia"}
+            if compute_device in {"cuda", "mtia", "npu"}
             else 0
         )
         for input_size, output_size, hbm_specific_size in zip(
diff --git a/torchrec/distributed/planner/stats.py b/torchrec/distributed/planner/stats.py
index 19fde330..0c88bff5 100644
--- a/torchrec/distributed/planner/stats.py
+++ b/torchrec/distributed/planner/stats.py
@@ -701,10 +701,7 @@ class EmbeddingStats(Stats):
             used_hbm_gb = bytes_to_gb(used_hbm[rank])
             used_hbm_ratio = (
                 used_hbm[rank] / ((1 - reserved_hbm_percent) * device.storage.hbm)
-                if (
-                    topology.compute_device == "cuda"
-                    or topology.compute_device == "mtia"
-                )
+                if topology.compute_device in {"cuda", "mtia", "npu"}
                 and ((1 - reserved_hbm_percent) * device.storage.hbm) != 0
                 else 0
             )
diff --git a/torchrec/distributed/planner/storage_reservations.py b/torchrec/distributed/planner/storage_reservations.py
index 6e7d3830..6fb9ee7b 100644
--- a/torchrec/distributed/planner/storage_reservations.py
+++ b/torchrec/distributed/planner/storage_reservations.py
@@ -73,7 +73,7 @@ def _reserve_dense_storage(
         dense_tensor_size = dense_tensor_estimate

     dense_tensor_storage = Storage(
-        hbm=dense_tensor_size if topology.compute_device in {"cuda", "mtia"} else 0,
+        hbm=dense_tensor_size if topology.compute_device in {"cuda", "mtia", "npu"} else 0,
         ddr=dense_tensor_size if topology.compute_device == "cpu" else 0,
     )

@@ -93,7 +93,7 @@ def _reserve_kjt_storage(
     kjt_size = math.ceil(sum(batch_inputs) * float(input_data_type_size)) * multiplier

     kjt_storage = Storage(
-        hbm=kjt_size if topology.compute_device in {"cuda", "mtia"} else 0,
+        hbm=kjt_size if topology.compute_device in {"cuda", "mtia", "npu"} else 0,
         ddr=kjt_size if topology.compute_device == "cpu" else 0,
     )

diff --git a/torchrec/distributed/planner/types.py b/torchrec/distributed/planner/types.py
index 4891bbc5..b2025cd9 100644
--- a/torchrec/distributed/planner/types.py
+++ b/torchrec/distributed/planner/types.py
@@ -309,13 +309,14 @@ class Topology:
             "cpu",
             "cuda",
             "mtia",
+            "npu"
         ], f"unsupported compute device {compute_device}"

         self._compute_device = compute_device
         self._world_size = world_size

         hbm_per_device = [0] * world_size
-        if self._compute_device == "cuda" or self._compute_device == "mtia":
+        if self._compute_device == "cuda" or self._compute_device == "npu":
             hbm_per_device = [hbm_cap if hbm_cap else HBM_CAP] * world_size
         ddr_cap_per_rank = [ddr_cap if ddr_cap else DDR_CAP] * world_size

diff --git a/torchrec/distributed/planner/utils.py b/torchrec/distributed/planner/utils.py
index 7d776305..1157894d 100644
--- a/torchrec/distributed/planner/utils.py
+++ b/torchrec/distributed/planner/utils.py
@@ -52,7 +52,7 @@ def placement(
     """

     param_device = compute_device
-    if compute_device in {"cuda", "mtia"}:
+    if compute_device in {"cuda", "mtia", "npu"}:
         param_device = torch.device(compute_device, rank % local_size)
     return f"rank:{rank}/{param_device}"

diff --git a/torchrec/distributed/quant_embedding.py b/torchrec/distributed/quant_embedding.py
index 0f91edc3..0fb121a7 100644
--- a/torchrec/distributed/quant_embedding.py
+++ b/torchrec/distributed/quant_embedding.py
@@ -215,7 +215,7 @@ def create_infer_embedding_sharding(
         get_device_from_sharding_infos(sharding_infos)
     )

-    if storage_device_type_from_sharding_infos in ["cuda", "mtia"]:
+    if storage_device_type_from_sharding_infos in ["cuda", "mtia", "npu"]:
         if sharding_type == ShardingType.TABLE_WISE.value:
             assert isinstance(
                 storage_device_type_from_sharding_infos, str
diff --git a/torchrec/distributed/quant_embedding_kernel.py b/torchrec/distributed/quant_embedding_kernel.py
index fe304df4..e94247fb 100644
--- a/torchrec/distributed/quant_embedding_kernel.py
+++ b/torchrec/distributed/quant_embedding_kernel.py
@@ -65,7 +65,7 @@ def _copy_config(
     config = copy.deepcopy(original)
     config.data_type = data_type
     for table in config.embedding_tables:
-        row_alignment = 16 if device.type == "cuda" else 1
+        row_alignment = 16 if (device.type == "cuda" or device.type == "npu") else 1
         table.local_cols = rounded_row_size_in_bytes(
             table.local_cols, sparse_type, row_alignment
         )
@@ -174,7 +174,7 @@ def _get_runtime_device(
                 )
                 for table in config.embedding_tables
             )
-            else torch.device("cuda")
+            else torch.device("npu" if torch.npu.is_available() else "cuda")
         )


@@ -264,7 +264,7 @@ class QuantBatchedEmbeddingBag(
         managed: List[EmbeddingLocation] = []
         is_virtual_table: bool = False
         for table in config.embedding_tables:
-            if device is not None and device.type == "cuda":
+            if device is not None and (device.type == "cuda" or device.type == "npu"):
                 managed.append(
                     compute_kernel_to_embedding_location(table.compute_kernel)
                 )
@@ -286,7 +286,11 @@ class QuantBatchedEmbeddingBag(
             device, config, shard_index
         )
         # 16 for CUDA, 1 for others like CPU and MTIA.
-        self._tbe_row_alignment: int = 16 if self._runtime_device.type == "cuda" else 1
+        self._tbe_row_alignment: int = (
+            16
+            if (self._runtime_device.type == "cuda" or self._runtime_device.type == "npu")
+            else 1
+        )
         embedding_specs = []
         for local_rows, local_cols, table, location in zip(
             self._local_rows,
@@ -528,7 +532,7 @@ class QuantBatchedEmbedding(
         is_virtual_table = False
         embedding_cache_mode = False
         for table in config.embedding_tables:
-            if device is not None and device.type == "cuda":
+            if device is not None and (device.type == "cuda" or device.type == "npu"):
                 managed.append(
                     compute_kernel_to_embedding_location(table.compute_kernel)
                 )
@@ -547,7 +551,11 @@ class QuantBatchedEmbedding(
             device, config, shard_index
         )
         # 16 for CUDA, 1 for others like CPU and MTIA.
-        self._tbe_row_alignment: int = 16 if self._runtime_device.type == "cuda" else 1
+        self._tbe_row_alignment: int = (
+            16
+            if (self._runtime_device.type == "cuda" or self._runtime_device.type == "npu")
+            else 1
+        )
         embedding_clazz = (
             KVEmbeddingInference
             if is_virtual_table
diff --git a/torchrec/distributed/quant_embeddingbag.py b/torchrec/distributed/quant_embeddingbag.py
index 1d774b37..71319224 100644
--- a/torchrec/distributed/quant_embeddingbag.py
+++ b/torchrec/distributed/quant_embeddingbag.py
@@ -449,7 +449,11 @@ class ShardedQuantFeatureProcessedEmbeddingBagCollection(
             device,
         )
         assert feature_processor is not None
-        device_type: str = self._device.type if self._device is not None else "cuda"
+        device_type: str = (
+            self._device.type
+            if self._device is not None else
+            ("npu" if torch.npu.is_available() else "cuda")
+        )
         self.feature_processors_per_rank: nn.ModuleList = torch.nn.ModuleList()
         feature_processor_device = None
         for _, param in feature_processor.named_parameters():
@@ -469,7 +473,7 @@ class ShardedQuantFeatureProcessedEmbeddingBagCollection(
                 )

         if isinstance(env, Dict):
-            expected_device_type = "cuda"
+            expected_device_type = "npu"
             for (
                 embedding_configs
             ) in self._sharding_type_device_group_to_sharding_infos.values():
@@ -501,7 +505,7 @@ class ShardedQuantFeatureProcessedEmbeddingBagCollection(
                         feature_processor_device,
                         (
                             torch.device(f"{device_type}:{i}")
-                            if device_type == "cuda"
+                            if (device_type == "cuda" or device_type == "npu")
                             else torch.device(f"{device_type}")
                         ),
                     )
diff --git a/torchrec/distributed/shard.py b/torchrec/distributed/shard.py
index 3c088042..d9bb1c66 100644
--- a/torchrec/distributed/shard.py
+++ b/torchrec/distributed/shard.py
@@ -123,6 +123,8 @@ def _shard(
         else:
             if torch.cuda.is_available():
                 device = torch.device(torch.cuda.current_device())
+            elif torch.npu.is_available():
+                device = torch.device(torch.npu.current_device())
             else:
                 device = torch.device("cpu")

@@ -246,6 +248,8 @@ def _shard_modules(  # noqa: C901
         else:
             if torch.cuda.is_available():
                 device = torch.device(torch.cuda.current_device())
+            elif torch.npu.is_available():
+                device = torch.device(torch.npu.current_device())
             else:
                 device = torch.device("cpu")

diff --git a/torchrec/distributed/sharding/dynamic_sharding.py b/torchrec/distributed/sharding/dynamic_sharding.py
index 075b1b41..92357d6f 100644
--- a/torchrec/distributed/sharding/dynamic_sharding.py
+++ b/torchrec/distributed/sharding/dynamic_sharding.py
@@ -699,7 +699,7 @@ def move_sharded_tensors_to_cpu(state_dict: Dict[str, Any]) -> Dict[str, Any]:
         if isinstance(item, ShardedTensor):
             # For ShardedTensor, move all local shards to CPU
             for shard in item.local_shards():
-                if shard.tensor.device.type == "cuda":
+                if shard.tensor.device.type == "cuda" or shard.tensor.device.type == "npu":
                     shard.tensor = shard.tensor.cpu()
             return item
         elif isinstance(item, dict):
@@ -716,7 +716,7 @@ def move_sharded_tensors_to_cpu(state_dict: Dict[str, Any]) -> Dict[str, Any]:
             return item

     processed_dict = _process_item(state_dict)
-    torch.cuda.empty_cache()
+    torch.npu.empty_cache()
     return processed_dict


diff --git a/torchrec/distributed/sharding/rw_pool_sharding.py b/torchrec/distributed/sharding/rw_pool_sharding.py
index e5bff5c7..8e05f920 100644
--- a/torchrec/distributed/sharding/rw_pool_sharding.py
+++ b/torchrec/distributed/sharding/rw_pool_sharding.py
@@ -243,6 +243,7 @@ class InferRwObjectPoolInputDist(torch.nn.Module):

         id_offsets = torch.ops.fbgemm.asynchronous_complete_cumsum(bucketized_lengths)
         dist_ids = []
+        device_type = "npu" if torch.npu.is_available() else "cuda"
         for rank in range(self._world_size):
             offset = id_offsets[rank]
             next_offset = id_offsets[rank + 1]
@@ -250,7 +251,9 @@ class InferRwObjectPoolInputDist(torch.nn.Module):
             dist_ids.append(
                 ids_for_rank
                 if self._device == torch.device("cpu")
-                else ids_for_rank.to(torch.device(f"cuda:{rank}"), non_blocking=True)
+                else ids_for_rank.to(
+                    torch.device(f"{device_type}:{rank}"), non_blocking=True
+                )
             )

         assert unbucketize_permute is not None, "unbucketize permute must not be None"
@@ -307,6 +310,7 @@ class InferRwObjectPoolInputDist(torch.nn.Module):
         unbucketize_values = values[index]
         dist_ids = []
         dist_values = []
+        device_type = "npu" if torch.npu.is_available() else "cuda"
         for rank in range(self._world_size):
             offset = id_offsets[rank]
             next_offset = id_offsets[rank + 1]
@@ -315,12 +319,12 @@ class InferRwObjectPoolInputDist(torch.nn.Module):
             dist_ids.append(
                 ids_for_rank
                 if self._device == torch.device("cpu")
-                else ids_for_rank.to(torch.device(f"cuda:{rank}"), non_blocking=True)
+                else ids_for_rank.to(torch.device(f"{device_type}:{rank}"), non_blocking=True)
             )
             dist_values.append(
                 values_for_rank
                 if self._device == torch.device("cpu")
-                else values_for_rank.to(torch.device(f"cuda:{rank}"), non_blocking=True)
+                else values_for_rank.to(torch.device(f"{device_type}:{rank}"), non_blocking=True)
             )

         return dist_ids, dist_values, unbucketize_permute
diff --git a/torchrec/distributed/sharding_plan.py b/torchrec/distributed/sharding_plan.py
index d70750d2..f301fcb2 100644
--- a/torchrec/distributed/sharding_plan.py
+++ b/torchrec/distributed/sharding_plan.py
@@ -75,7 +75,7 @@ def placement(
     local_size: int,
 ) -> str:
     param_device = compute_device
-    if compute_device in {"cuda", "mtia"}:
+    if compute_device in {"cuda", "mtia", "npu"}:
         param_device = torch.device(compute_device, rank % local_size)
     return f"rank:{rank}/{param_device}"

@@ -350,7 +350,7 @@ def _get_compute_kernel(
         compute_kernels += [
             EmbeddingComputeKernel.FUSED.value,
         ]
-    if device_type in {"cuda"}:
+    if device_type in {"cuda", "npu"}:
         compute_kernels += [
             EmbeddingComputeKernel.FUSED_UVM.value,
             EmbeddingComputeKernel.FUSED_UVM_CACHING.value,
@@ -631,7 +631,7 @@ def row_wise(
                     device_type = str(sizes_placement[1])
                     placements.append(placement_helper(device_type, index, i))

-                if device_type == "cuda":
+                if device_type == "cuda" or device_type == "npu":
                     index += 1

         return _get_parameter_sharding(
diff --git a/torchrec/distributed/tensor_pool.py b/torchrec/distributed/tensor_pool.py
index 21d44b56..7c5a694d 100644
--- a/torchrec/distributed/tensor_pool.py
+++ b/torchrec/distributed/tensor_pool.py
@@ -346,7 +346,7 @@ class ShardedInferenceTensorPool(
         self._dtype = dtype
         self._sharding_env = env
         self._world_size = env.world_size
-        self._device = device or torch.device("cuda")
+        self._device = device or torch.device("npu" if torch.npu.is_available() else "cuda")
         self._sharding_plan = plan

         self._rank = env.rank
@@ -384,7 +384,7 @@ class ShardedInferenceTensorPool(
             shard_device = (
                 torch.device("cpu")
                 if device == torch.device("cpu")
-                else torch.device("cuda", rank)
+                else torch.device("npu" if torch.npu.is_available() else "cuda", rank)
             )

             self._local_shard_pools.append(
diff --git a/torchrec/distributed/train_pipeline/train_pipelines.py b/torchrec/distributed/train_pipeline/train_pipelines.py
index 2293a087..977e2d7d 100644
--- a/torchrec/distributed/train_pipeline/train_pipelines.py
+++ b/torchrec/distributed/train_pipeline/train_pipelines.py
@@ -212,7 +212,7 @@ class TrainPipelineBase(TrainPipeline[In, Out]):
         self._device = device
         self._memcpy_stream: Optional[torch.Stream] = (
             torch.get_device_module(device).Stream()
-            if device.type in ["cuda", "mtia"]
+            if device.type in ["cuda", "mtia", "npu"]
             else None
         )
         self._inplace_copy_batch_to_gpu = inplace_copy_batch_to_gpu
@@ -223,7 +223,7 @@ class TrainPipelineBase(TrainPipeline[In, Out]):
         # pyre-ignore
         self._stream_context = (
             torch.get_device_module(self._device).stream
-            if self._device.type in ["cuda", "mtia"]
+            if self._device.type in ["cuda", "mtia", "npu"]
             else torch.cuda.stream
         )
         self._cur_batch: Optional[In] = None
@@ -520,22 +520,26 @@ class TrainPipelineSparseDist(TrainPipeline[In, Out]):
             assert (
                 not is_torchdynamo_compiling()
             ), "Train Pipelines rely on cuda streams, which is not supported by Dynamo"
+        if device.type == "npu":
+            assert (
+                not is_torchdynamo_compiling()
+            ), "Train Pipelines rely on npu streams, which is not supported by Dynamo"

         # pyre-ignore
         self._stream_context = (
             torch.get_device_module(self._device).stream
-            if self._device.type in ["cuda", "mtia"]
+            if self._device.type in ["cuda", "mtia", "npu"]
             else torch.cuda.stream
         )

         self._memcpy_stream: Optional[torch.Stream] = (
             (torch.get_device_module(device).Stream(priority=-1))
-            if device.type in ["cuda", "mtia"]
+            if device.type in ["cuda", "mtia", "npu"]
             else None
         )
         self._data_dist_stream: Optional[torch.Stream] = (
             (torch.get_device_module(device).Stream(priority=-1))
-            if device.type in ["cuda", "mtia"]
+            if device.type in ["cuda", "mtia", "npu"]
             else None
         )

@@ -1099,7 +1103,7 @@ class TrainPipelineFusedSparseDist(TrainPipelineSparseDist[In, Out]):
         if emb_lookup_stream == "new":
             self._emb_lookup_stream: Optional[torch.Stream] = (
                 (torch.get_device_module(device).Stream())
-                if device.type in ["cuda", "mtia"]
+                if device.type in ["cuda", "mtia", "npu"]
                 else None
             )
         elif emb_lookup_stream == "current":
@@ -1651,12 +1655,12 @@ class PrefetchTrainPipelineSparseDist(TrainPipelineSparseDist[In, Out]):
         self._context = PrefetchTrainPipelineContext(version=0)
         self._prefetch_stream: Optional[torch.Stream] = (
             (torch.get_device_module(device).Stream())
-            if self._device.type in ["cuda", "mtia"]
+            if self._device.type in ["cuda", "mtia", "npu"]
             else None
         )
         self._default_stream: Optional[torch.Stream] = (
             (torch.get_device_module(self._device).Stream())
-            if self._device.type in ["cuda", "mtia"]
+            if self._device.type in ["cuda", "mtia", "npu"]
             else None
         )
         self._batch_ip3: Optional[In] = None
@@ -1974,7 +1978,7 @@ class StagedTrainPipeline(TrainPipeline[In, Optional[StageOut]]):
         # pyre-ignore
         self._stream_context = (
             torch.get_device_module(self._compute_stream.device).stream
-            if self._compute_stream.device.type in ["cuda", "mtia"]
+            if self._compute_stream.device.type in ["cuda", "mtia", "npu"]
             else torch.cuda.stream
         )

diff --git a/torchrec/distributed/train_pipeline/utils.py b/torchrec/distributed/train_pipeline/utils.py
index 7842ff04..d60d7b61 100644
--- a/torchrec/distributed/train_pipeline/utils.py
+++ b/torchrec/distributed/train_pipeline/utils.py
@@ -508,7 +508,7 @@ class DataLoadingThread(Thread, Generic[In]):
         if memcpy_stream is None:
             self._memcpy_stream: Optional[torch.Stream] = (
                 torch.get_device_module(device).Stream(priority=memcpy_stream_priority)
-                if device.type in ["cuda", "mtia"]
+                if device.type in ["cuda", "mtia", "npu"]
                 else None
             )
         else:
@@ -522,6 +522,9 @@ class DataLoadingThread(Thread, Generic[In]):
         if self._device.type == "cuda" and torch.cuda.is_available():
             # set the current device the same as the one used in the main thread
             torch.cuda.set_device(self._device)
+        elif self._device.type == "npu" and torch.npu.is_available():
+            # set the current device the same as the one used in the main thread
+            torch.npu.set_device(self._device)
         elif self._device.type == "mtia" and torch.mtia.is_available():
             # set the current device the same as the one used in the main thread
             torch.mtia.set_device(self._device)
diff --git a/torchrec/distributed/types.py b/torchrec/distributed/types.py
index f81c9289..ca170e10 100644
--- a/torchrec/distributed/types.py
+++ b/torchrec/distributed/types.py
@@ -477,7 +477,7 @@ class DeviceToHostTensorAwaitable(LazyAwaitable[torch.Tensor]):
         self._tensor: torch.Tensor = tensor_on_device.to("cpu", non_blocking=True)

         # cuda event to record the completion of the copy
-        self._event = torch.cuda.Event()
+        self._event = torch.npu.Event()
         self._event.record()

     def _wait_impl(self) -> torch.Tensor:
@@ -1410,9 +1410,10 @@ class ModuleSharder(abc.ABC, Generic[M]):
         compute kernel.
         """

-        assert compute_device_type in {"cuda", "cpu", "mtia"}
+        assert compute_device_type in {"cuda", "cpu", "mtia", "npu"}
         storage_map = {
             "cuda": ParameterStorage.HBM,
+            "npu": ParameterStorage.HBM,
             "cpu": ParameterStorage.DDR,
             "mtia": ParameterStorage.HBM,
         }
diff --git a/torchrec/distributed/utils.py b/torchrec/distributed/utils.py
index a05456d0..e10f8dc9 100644
--- a/torchrec/distributed/utils.py
+++ b/torchrec/distributed/utils.py
@@ -57,6 +57,8 @@ def get_device_type() -> str:
         device_type = "cuda"
     elif torch.mtia.is_available():
         device_type = "mtia"
+    elif torch.npu.is_available():
+        device_type = "npu"
     else:
         device_type = "cpu"
     return device_type
@@ -230,14 +232,20 @@ class sharded_model_copy:

         torch.Tensor.__deepcopy__ = _copy_or_not
         torch.nn.Parameter.__deepcopy__ = _param_copy
-        # pyre-fixme[16]: `Type` has no attribute `__deepcopy__`.
-        torch._C._distributed_c10d.ProcessGroupNCCL.__deepcopy__ = _no_copy
+        if torch.npu.is_available():
+            # pyre-fixme[16]: `Type` has no attribute `__deepcopy__`.
+            torch._C._distributed_c10d.ProcessGroup.__deepcopy__ = _no_copy
+            # pyre-ignore [16]
+            torch.npu.streams.Stream.__deepcopy__ = _no_copy
+        else:
+            # pyre-fixme[16]: `Type` has no attribute `__deepcopy__`.
+            torch._C._distributed_c10d.ProcessGroupNCCL.__deepcopy__ = _no_copy
+            # pyre-ignore [16]
+            torch.cuda.streams.Stream.__deepcopy__ = _no_copy
         # pyre-fixme[16]: `Type` has no attribute `__deepcopy__`.
         torch._C._distributed_c10d.ProcessGroupGloo.__deepcopy__ = _no_copy
         # pyre-fixme[16]: `Type` has no attribute `__deepcopy__`.
         torch._C._distributed_c10d.Work.__deepcopy__ = _no_copy
-        # pyre-ignore [16]
-        torch.cuda.streams.Stream.__deepcopy__ = _no_copy

     # pyre-ignore [2]
     def __exit__(self, exc_type, exc_val, exc_tb) -> None:
@@ -245,14 +253,20 @@ class sharded_model_copy:
         torch.Tensor.__deepcopy__ = self.t_copy_save_
         # pyre-ignore [16]
         torch.nn.Parameter.__deepcopy__ = self.p_copy_save_
-        # pyre-fixme[16]: `Type` has no attribute `__deepcopy__`.
-        torch._C._distributed_c10d.ProcessGroupNCCL.__deepcopy__ = None
+        if torch.npu.is_available():
+            # pyre-fixme[16]: `Type` has no attribute `__deepcopy__`.
+            torch._C._distributed_c10d.ProcessGroup.__deepcopy__ = None
+            # pyre-ignore [16]
+            torch.npu.streams.Stream.__deepcopy__ = None
+        else:
+            # pyre-fixme[16]: `Type` has no attribute `__deepcopy__`.
+            torch._C._distributed_c10d.ProcessGroupNCCL.__deepcopy__ = None
+            # pyre-ignore [16]
+            torch.cuda.streams.Stream.__deepcopy__ = None
         # pyre-fixme[16]: `Type` has no attribute `__deepcopy__`.
         torch._C._distributed_c10d.ProcessGroupGloo.__deepcopy__ = None
         # pyre-fixme[16]: `Type` has no attribute `__deepcopy__`.
         torch._C._distributed_c10d.Work.__deepcopy__ = None
-        # pyre-ignore [16]
-        torch.cuda.streams.Stream.__deepcopy__ = None


 def copy_to_device(
diff --git a/torchrec/inference/modules.py b/torchrec/inference/modules.py
index 5458cc4f..eae2339d 100644
--- a/torchrec/inference/modules.py
+++ b/torchrec/inference/modules.py
@@ -308,7 +308,10 @@ class PredictModule(nn.Module):
     # pyre-fixme[3]
     def forward(self, batch: Dict[str, torch.Tensor]) -> Any:
         if self._device is None:
-            self._device = torch.device("cuda", torch.cuda.current_device())
+            if torch.npu.is_available():
+                self._device = torch.device("npu", torch.npu.current_device())
+            else:
+                self._device = torch.device("cuda", torch.cuda.current_device())
         with torch.inference_mode():
             return self.predict_forward(batch)

@@ -575,6 +578,10 @@ def shard_quant_model(
         hbm_cap = torch.cuda.get_device_properties(
             f"cuda:{torch.cuda.current_device()}"
         ).total_memory
+    elif torch.npu.is_available() and compute_device == "npu":
+        hbm_cap = torch.npu.get_device_properties(
+            f"npu:{torch.npu.current_device()}"
+        ).total_memory
     else:
         hbm_cap = None

diff --git a/torchrec/modules/fused_embedding_modules.py b/torchrec/modules/fused_embedding_modules.py
index 1464118d..9eb316b7 100644
--- a/torchrec/modules/fused_embedding_modules.py
+++ b/torchrec/modules/fused_embedding_modules.py
@@ -148,6 +148,8 @@ class _BatchedFusedEmbeddingLookups(nn.Module, FusedOptimizerModule):
         compute_device = ComputeDevice.CPU
         if device.type == "cuda":
             compute_device = ComputeDevice.CUDA
+        elif device.type == "npu":
+            compute_device = ComputeDevice.NPU

         self._emb_module: SplitTableBatchedEmbeddingBagsCodegen = (
             SplitTableBatchedEmbeddingBagsCodegen(
@@ -375,8 +377,9 @@ class FusedEmbeddingBagCollection(
         ]:
             assert device is not None and device.type in [
                 "cuda",
+                "npu",
                 "meta",
-            ], f"Using location={location} requires device=cuda or meta"
+            ], f"Using location={location} requires device=cuda or npu or meta"

         if device is None:
             device = torch.device("cpu")
@@ -386,6 +389,8 @@ class FusedEmbeddingBagCollection(
                 location = EmbeddingLocation.HOST
             elif device.type == "cuda":
                 location = EmbeddingLocation.DEVICE
+            elif device.type == "npu":
+                location = EmbeddingLocation.DEVICE
             else:
                 raise ValueError("EmbeddingLocation could not be set")

@@ -612,22 +617,26 @@ class FusedEmbeddingCollection(EmbeddingCollectionInterface, FusedOptimizerModul
         ]:
             assert device is not None and device.type in [
                 "cuda",
+                "npu",
                 "meta",
-            ], f"Using location={location} requires device=cuda or meta"
+            ], f"Using location={location} requires device=cuda or npu or meta"

         if device is None:
             device = torch.device("cpu")

         assert device.type in [
             "cuda",
+            "npu",
             "meta",
-        ], "FusedEmbeddingCollection is only supported for device in [CUDA, meta] currently. There are plans to support device=CPU."
+        ], "FusedEmbeddingCollection is only supported for device in [CUDA, NPU, meta] currently. There are plans to support device=CPU."

         if location is None:
             if device.type in ["cpu", "meta"]:
                 location = EmbeddingLocation.HOST
             elif device.type == "cuda":
                 location = EmbeddingLocation.DEVICE
+            elif device.type == "npu":
+                location = EmbeddingLocation.DEVICE
             else:
                 raise ValueError("EmbeddingLocation could not be set")

diff --git a/torchrec/modules/itep_modules.py b/torchrec/modules/itep_modules.py
index 28335420..7a4a0d4b 100644
--- a/torchrec/modules/itep_modules.py
+++ b/torchrec/modules/itep_modules.py
@@ -314,7 +314,7 @@ class GenericITEPModule(nn.Module):
                             ), f"Device of table {table}: {table.local_metadata.placement.device()} does not match existing device: {self.current_device}"

         if self.current_device is None:
-            self.current_device = torch.device("cuda")
+            self.current_device = torch.device("npu" if torch.npu.is_available() else "cuda")

         self.reversed_feature_table_map: Dict[int, str] = {
             idx: feature_name for feature_name, idx in self.feature_table_map.items()
diff --git a/torchrec/modules/object_pool_lookups.py b/torchrec/modules/object_pool_lookups.py
index 0368a58f..d83ca808 100644
--- a/torchrec/modules/object_pool_lookups.py
+++ b/torchrec/modules/object_pool_lookups.py
@@ -321,13 +321,14 @@ class UVMCachingInt64Lookup(KeyedJaggedTensorPoolLookup):
             [self._bit_dims], dtype=torch.int32, device=self._device
         )

+        compute_device = ComputeDevice.NPU if torch.npu.is_available() else ComputeDevice.CUDA
         self._tbe = SplitTableBatchedEmbeddingBagsCodegen(
             embedding_specs=[
                 (
                     pool_size,
                     2 * self._bit_dims,
                     EmbeddingLocation.MANAGED,
-                    ComputeDevice.CUDA,
+                    compute_device,
                 ),
             ],
             pooling_mode=PoolingMode.NONE,
@@ -349,7 +350,7 @@ class UVMCachingInt64Lookup(KeyedJaggedTensorPoolLookup):
                             if device != torch.device("meta")
                             else EmbeddingLocation.MANAGED
                         ),
-                        ComputeDevice.CUDA,
+                        compute_device,
                     ),
                 ],
                 pooling_mode=PoolingMode.NONE,
@@ -464,6 +465,7 @@ class UVMCachingInt32Lookup(KeyedJaggedTensorPoolLookup):
             [self._bit_dims], dtype=torch.int32, device=self._device
         )

+        compute_device = ComputeDevice.NPU if torch.npu.is_available() else ComputeDevice.CUDA
         self._tbe = SplitTableBatchedEmbeddingBagsCodegen(
             embedding_specs=[
                 (
@@ -474,7 +476,7 @@ class UVMCachingInt32Lookup(KeyedJaggedTensorPoolLookup):
                         if self._device.type != "meta"
                         else EmbeddingLocation.DEVICE
                     ),
-                    ComputeDevice.CUDA,
+                    compute_device,
                 ),
             ],
             pooling_mode=PoolingMode.NONE,
@@ -497,7 +499,7 @@ class UVMCachingInt32Lookup(KeyedJaggedTensorPoolLookup):
                             if self._device.type != "meta"
                             else EmbeddingLocation.DEVICE
                         ),
-                        ComputeDevice.CUDA,
+                        compute_device,
                     ),
                 ],
                 pooling_mode=PoolingMode.NONE,
@@ -725,7 +727,7 @@ class UVMCachingFloatLookup(TensorPoolLookup):
                         if self._device.type != "meta"
                         else EmbeddingLocation.DEVICE
                     ),
-                    ComputeDevice.CUDA,
+                    ComputeDevice.NPU if torch.npu.is_available() else ComputeDevice.CUDA,
                 ),
             ],
             pooling_mode=PoolingMode.NONE,
diff --git a/torchrec/modules/utils.py b/torchrec/modules/utils.py
index c4fc7c0e..8e2ac24a 100644
--- a/torchrec/modules/utils.py
+++ b/torchrec/modules/utils.py
@@ -179,7 +179,7 @@ def _permute_tensor_by_segments(
     NOTE:
         `keyed_jagged_index_select_dim1` is only supported for CUDA.
     """
-    if tensor.device.type == "cuda":
+    if tensor.device.type == "cuda" or tensor.device.type == "npu":
         output = torch.ops.fbgemm.keyed_jagged_index_select_dim1(
             values=tensor,
             lengths=segment_sizes,
diff --git a/torchrec/sparse/jagged_tensor.py b/torchrec/sparse/jagged_tensor.py
index df73074c..e1efac17 100644
--- a/torchrec/sparse/jagged_tensor.py
+++ b/torchrec/sparse/jagged_tensor.py
@@ -59,7 +59,7 @@ def _pin_and_move(tensor: torch.Tensor, device: torch.device) -> torch.Tensor:

     return (
         tensor.pin_memory().to(device=device, non_blocking=True)
-        if device.type == "cuda" and tensor.device.type == "cpu"
+        if (device.type == "cuda" or device.type == "npu") and tensor.device.type == "cpu"
         else tensor.to(device=device, non_blocking=True)
     )

@@ -447,7 +447,7 @@ def _permute_tensor_by_segments(
     NOTE:
         `keyed_jagged_index_select_dim1` is only supported for CUDA.
     """
-    if tensor.device.type == "cuda" and (
+    if (tensor.device.type == "cuda" or tensor.device.type == "npu") and (
         weights is None or weights.dtype == torch.float32
     ):
         output = torch.ops.fbgemm.keyed_jagged_index_select_dim1(
diff --git a/version.txt b/version.txt
index bc80560f..16eaa17e 100644
--- a/version.txt
+++ b/version.txt
@@ -1 +1 @@
-1.5.0
+1.5.0+npu
