diff --git a/examples/hstu/benchmark/inference_benchmark_random_2k.py b/examples/hstu/benchmark/inference_benchmark_random_2k.py
index 064ccb0..6aa8614 100644
--- a/examples/hstu/benchmark/inference_benchmark_random_2k.py
+++ b/examples/hstu/benchmark/inference_benchmark_random_2k.py
@@ -47,7 +47,8 @@ use_cudaGraph = int(os.getenv("USR_CUDA_GRAPH")) == 1
 dataset_path = os.getenv("DATASET_PATH")
 model_path = os.getenv("MODEL_PATH")
 folder_name = os.getenv("PROF_SAVE_PATH")
-is_save_input_and_output = int(os.getenv("SAVE_INPUT_AND_PUTPUT")) == 1
+is_save_input_and_output = int(os.getenv("SAVE_INPUT_AND_OUTPUT", "0")) == 1
+is_save_model = int(os.getenv("SAVE_MODEL", "0")) == 1
 is_use_npu_input = int(os.getenv("USE_NPU_INPUT")) == 1
 batch_size = int(os.getenv("BATCH_SIZE", "1"))
 layer_num = int(os.getenv("LAYER_NUM", "1"))
@@ -149,8 +150,8 @@ def run_ranking_gr_inference():
 
         if use_random_parm:
             for param in model_predict.parameters():
-                torch.nn.init.uniform_(param, a=-parm_value, b=parm_value)
-            if is_save_input_and_output:
+                torch.nn.init.normal_(param,mean=0, std=parm_value)
+            if is_save_model:
                 cpu_state_dict = {k: v.cpu() for k, v in model_predict.state_dict().items()}
                 torch.save(cpu_state_dict, model_path)  # 保存模型权重
         else:
@@ -213,103 +214,111 @@ def run_ranking_gr_inference():
 
         for item in infer_list * run_num:
             result = model_predict.forward(item[0], item[1])
-            # print([model output is],result)
             if is_save_input_and_output:
                 torch.save(result.cpu(), "input_output/gpu_output.pt")
 
         print("性能测试-运行开始！")
-        time_arr = []
-        with torch.profiler.profile(
-                activities=[ProfilerActivity.CUDA, ProfilerActivity.CPU],  # 同时分析 CPU 和 CUDA
-                record_shapes=True,  # 记录张量形状
-        ) as prof:
-            with torch.no_grad():
-                torch.cuda.synchronize()
-                for item in infer_list * run_num:
-                    model_predict.forward(item[0], item[1])  # 推理
-                    torch.cuda.synchronize()
-                    prof.step()
-
-        print("收集端到端时间")
-        with torch.no_grad():
-            for item in infer_list * run_num:
-                torch.cuda.synchronize()
-                start_time = time.perf_counter()
-                model_predict.forward(item[0], item[1])  # 推理
-                torch.cuda.synchronize()
-                duration = (time.perf_counter() - start_time) * 1_000_000
-                time_arr.append(duration)
-                print(f"[E2E] time: {duration:.2f} us")
-        report = {'Batch_size': max_batch_size,
-                  'model_name': "GR"}
-
-        time_arr.sort()
-        tail_latency = round(time_arr[int(len(time_arr) * 0.99)] / 1000, 6)
-        p90_latency = round(time_arr[int(len(time_arr) * 0.90)] / 1000, 6)
-        avg_latency = round(sum(time_arr) / len(time_arr) / 1000, 6)
-        qps = int(1000.0 * max_batch_size / avg_latency)
-
-        report["QPS"] = qps
-        report["AVG Latency"] = avg_latency
-        report["P99 Latency"] = tail_latency
-        report["P90 Latency"] = p90_latency
-        print(report)
         time_str = datetime.datetime.now().strftime("%Y%m%d_%H%M%S_%f")
 
         if not os.path.exists(folder_name):
             os.makedirs(folder_name)
 
-        with open(os.path.join(folder_name, f'report_bs{max_batch_size}.json'), "w", encoding="utf-8") as f:
-            json.dump(
-                report,
-                f,
-                ensure_ascii=False,  # 保留非ASCII字符（如中文）
-                indent=4  # 格式化输出，缩进4个空格，增强可读性
-            )
+        if os.getenv('USE_PROFILING', '0') != '0':
+            with torch.profiler.profile(
+                    activities=[ProfilerActivity.CUDA, ProfilerActivity.CPU],  # 同时分析 CPU 和 CUDA
+                    record_shapes=True,  # 记录张量形状
+            ) as prof:
+                steps = min(10, run_num)
+                with torch.no_grad():
+                    torch.cuda.synchronize()
+                    for item in infer_list * steps:
+                        model_predict.forward(item[0], item[1])  # 推理
+                        torch.cuda.synchronize()
+                        prof.step()
+            output_trace_path = os.path.join(folder_name, f'{time_str}_trace.json')
+            prof.export_chrome_trace(output_trace_path)
+
+            # 3. 获取聚合后的性能数据
+            # key_averages() 会按操作名分组聚合事件
+            key_averages_events = prof.key_averages()
+            if key_averages_events:
+                print("可用的 Event 属性列表:")
+                print(dir(key_averages_events[0]))
+
+            # 4. 将 profiler 结果转换为 Pandas DataFrame
+            # 每个事件都有如 name, cpu_time_total, cuda_time_total 等属性
+            # 我们可以用 map(vars, ...) 快速地将每个事件对象转换为字典
+            profiler_data = []
+            for event in key_averages_events:
+                profiler_data.append({
+                    'Name': event.key,
+                    'CPU Time Total (us)': event.cpu_time_total,
+                    'CUDA Time Total (us)': event.device_time_total,  # <-- 修正
+                    'CPU Time Self (us)': event.self_cpu_time_total,
+                    'CUDA Time Self (us)': event.self_device_time_total,  # <-- 修正
+                    'CPU Memory Self (B)': event.self_cpu_memory_usage,
+                    'CUDA Memory Self (B)': event.self_device_memory_usage,  # <-- 修正
+                    'Number of Calls': event.count,
+                })
+
+            df = pd.DataFrame(profiler_data)
+
+            # 5. (可选) 对数据进行排序，方便分析
+            # 按 CUDA 总时间降序排列
+            df_sorted = df.sort_values(by='CUDA Time Total (us)', ascending=False)
+
+            # 6. 导出到 Excel 文件
+
+            output_excel_path = os.path.join(folder_name, f'{time_str}_profiler_results.xlsx')
+            df_sorted.to_excel(output_excel_path, index=False)
+
+            print(f"Profiler results have been exported to {output_excel_path}")
+
+        if os.getenv('USE_E2E', '1') != '0':
+            print("收集端到端时间")
+            all_time = []
 
+            with torch.no_grad():
+                for item in infer_list * run_num:
+                    torch.cuda.synchronize()
+                    start_time = time.perf_counter()
+                    model_predict.forward(item[0], item[1])  # 推理
+                    torch.cuda.synchronize()
+                    duration = (time.perf_counter() - start_time) * 1_000_000
+                    all_time.append(duration)
+                    if (len(all_time) % 100 == 0):
+                        print(f"[E2E] time: {duration:.2f} us")
+
+            report = {'Batch_size': max_batch_size,
+                    'model_name': "GR"}
+
+            all_time.sort()
+            p999_latency = round(all_time[int(len(all_time) * 0.999)] / 1000, 6)
+            tail_latency = round(all_time[int(len(all_time) * 0.99)] / 1000, 6)
+            p95_latency = round(all_time[int(len(all_time) * 0.95)] / 1000, 6)
+            p90_latency = round(all_time[int(len(all_time) * 0.90)] / 1000, 6)
+            avg_latency = round(sum(all_time) / len(all_time) / 1000, 6)
+            qps = int(1000.0 * max_batch_size / avg_latency)
+
+            report["QPS"] = qps
+            report["AVG Latency"] = avg_latency
+            report["P999 Latency"] = p999_latency
+            report["P99 Latency"] = tail_latency
+            report["P95 Latency"] = p95_latency
+            report["P90 Latency"] = p90_latency
+        
+            print(report)
+            with open(os.path.join(folder_name, f'report_bs{max_batch_size}.json'), "w", encoding="utf-8") as f:
+                json.dump(
+                    report,
+                    f,
+                    ensure_ascii=False,  # 保留非ASCII字符（如中文）
+                    indent=4  # 格式化输出，缩进4个空格，增强可读性
+                )
+        print(f"计算过程中显存峰值：{torch.cuda.max_memory_allocated() / 1024 ** 2:.2f} MB")
         print("性能分析完成！")
         print("-" * 30)
 
-        print(prof.key_averages())
-        output_trace_path = os.path.join(folder_name, f'{time_str}_trace.json')
-        prof.export_chrome_trace(output_trace_path)
-
-        # 3. 获取聚合后的性能数据
-        # key_averages() 会按操作名分组聚合事件
-        key_averages_events = prof.key_averages()
-        if key_averages_events:
-            print("可用的 Event 属性列表:")
-            print(dir(key_averages_events[0]))
-
-        # 4. 将 profiler 结果转换为 Pandas DataFrame
-        # 每个事件都有如 name, cpu_time_total, cuda_time_total 等属性
-        # 我们可以用 map(vars, ...) 快速地将每个事件对象转换为字典
-        profiler_data = []
-        for event in key_averages_events:
-            profiler_data.append({
-                'Name': event.key,
-                'CPU Time Total (us)': event.cpu_time_total,
-                'CUDA Time Total (us)': event.device_time_total,  # <-- 修正
-                'CPU Time Self (us)': event.self_cpu_time_total,
-                'CUDA Time Self (us)': event.self_device_time_total,  # <-- 修正
-                'CPU Memory Self (B)': event.self_cpu_memory_usage,
-                'CUDA Memory Self (B)': event.self_device_memory_usage,  # <-- 修正
-                'Number of Calls': event.count,
-            })
-
-        df = pd.DataFrame(profiler_data)
-
-        # 5. (可选) 对数据进行排序，方便分析
-        # 按 CUDA 总时间降序排列
-        df_sorted = df.sort_values(by='CUDA Time Total (us)', ascending=False)
-
-        # 6. 导出到 Excel 文件
-
-        output_excel_path = os.path.join(folder_name, f'{time_str}_profiler_results.xlsx')
-        df_sorted.to_excel(output_excel_path, index=False)
-
-        print(f"Profiler results have been exported to {output_excel_path}")
-
 
 if __name__ == "__main__":
     torch.manual_seed(42)
diff --git a/examples/hstu/modules/hstu_block_inference.py b/examples/hstu/modules/hstu_block_inference.py
index 74ec9c2..c79c713 100644
--- a/examples/hstu/modules/hstu_block_inference.py
+++ b/examples/hstu/modules/hstu_block_inference.py
@@ -54,6 +54,7 @@ class HSTUBlockInference(torch.nn.Module):
             ]
         )
         self._hstu_graph: Optional[Dict[int, Any]] = None  # type: ignore
+        self._compile_predict = None
 
     def hstu_preprocess(
             self, embeddings: Dict[str, JaggedTensor], batch: Batch
@@ -139,19 +140,31 @@ class HSTUBlockInference(torch.nn.Module):
             use_cudagraph: bool = True,
     ) -> torch.Tensor:
         if self._hstu_graph is None or not use_cudagraph:
-            hidden_data = hidden_states
-            for hstu_layer in self._attention_layers:
-                hidden_data = hstu_layer.forward_naive(
-                    batch_size, num_tokens, hidden_data, jd
+            if os.getenv("USE_COMPILE", "0") == "1" and self._compile_predict is None:
+                self._compile_predict = torch.compile(
+                    self.predict_nograph, backend="inductor", dynamic=False
                 )
-            return hidden_data
+            predict_func = self.predict_nograph if self._compile_predict is None else self._compile_predict
+            return predict_func(batch_size, num_tokens, hidden_states, jd)
         else:
-
-            # OK了吗print(' YZ: we are inside the hstu_block_inference.py ')
             return self.predict_cudagraph(
                 batch_size, num_tokens, hidden_states
             )
 
+    def predict_nograph(
+        self,
+        batch_size: int,
+        num_tokens: int,
+        hidden_states: torch.Tensor,
+        jd: JaggedData,
+    )-> torch.Tensor:
+        hidden_data = hidden_states
+        for hstu_layer in self._attention_layers:
+            hidden_data = hstu_layer.forward_naive(
+                batch_size, num_tokens, hidden_data, jd
+            )
+        return hidden_data
+
     def predict_naive(
             self,
             batch_size: int,
diff --git a/examples/hstu/modules/paged_hstu_infer_layer.py b/examples/hstu/modules/paged_hstu_infer_layer.py
index b7b740e..d92d577 100644
--- a/examples/hstu/modules/paged_hstu_infer_layer.py
+++ b/examples/hstu/modules/paged_hstu_infer_layer.py
@@ -51,12 +51,13 @@ class SparseMOE(torch.nn.Module):
             emb_dim,
             expert_number,
             top_k,
+            hidden_size,
             dtype
         ):
         super().__init__()
         self.expert_number = expert_number
         self.experts = torch.nn.ModuleList([
-                FFN_npu_swiglu(emb_dim, emb_dim * 4, dtype) for _ in range(expert_number)])
+                FFN_npu_swiglu(emb_dim, hidden_size, dtype) for _ in range(expert_number)])
 
         self.router = MOERouter(emb_dim, expert_number, top_k, dtype)
         self.w1,self.w2,self.w3 = self.create_expert_weights(self.experts)
@@ -476,11 +477,12 @@ class PagedHSTUInferLayer(torch.nn.Module):
                 self.feed_forward = torch.nn.ModuleList([
                     FFN_npu_swiglu(self._embedding_dim, self._embedding_dim * 4, dtype) for _ in
                     range(self.ffn_number)])
-        if self.use_moe:
+        if self.use_moe and self.layer_idx > 1:
             self.sparse_moe_linear = SparseMOE(
                 emb_dim=self._embedding_dim,
                 expert_number=int(os.getenv("MOE_EXPERT_TOTAL_NUM")),
                 top_k=int(os.getenv("MOE_EXPERT_ACT_NUM")),
+                hidden_size=int(os.getenv("MOE_HIDDEN_SIZE")),
                 dtype=dtype
             )
 
@@ -493,7 +495,7 @@ class PagedHSTUInferLayer(torch.nn.Module):
             )
 
     def post_init(self):
-        if self.use_moe:
+        if self.use_moe and self.layer_idx > 1:
             self.sparse_moe_linear.load_parm_post_init()
         if self.use_rab:
             self.relative_bias.load_parm_post_init()
@@ -554,11 +556,11 @@ class PagedHSTUInferLayer(torch.nn.Module):
                 max_seqlen_q=self._max_seqlen,
                 max_seqlen_k=self._max_seqlen,
                 num_contexts= None,
-                num_targets= None,
+                num_targets=jd.num_candidates,
                 target_group_size=1,
                 window_size=(-1, 0),
-                alpha=self._alpha*self._max_seqlen,
-                rab= rab,
+                alpha=self._alpha,
+                rab=rab,
                 has_drab=False,
                 is_delta_q=False,
             )
@@ -584,7 +586,7 @@ class PagedHSTUInferLayer(torch.nn.Module):
             layer_output = self._linear_proj(parallel_input)
 
 
-        if self.use_moe :
+        if self.use_moe and self.layer_idx > 1:
             layer_output = self.sparse_moe_linear.forward(layer_output) if self.use_moe_op else self.sparse_moe_linear.forward_naive(layer_output)
 
 
@@ -665,11 +667,11 @@ class PagedHSTUInferLayer(torch.nn.Module):
                 max_seqlen_q=self._max_seqlen,
                 max_seqlen_k=self._max_seqlen,
                 num_contexts= None,
-                num_targets= None,
+                num_targets=jd.num_candidates,
                 target_group_size=1,
                 window_size=(-1, 0),
-                alpha=self._alpha*self._max_seqlen,
-                rab= rab,
+                alpha=self._alpha,
+                rab=rab,
                 has_drab=False,
                 is_delta_q=False,
             )
@@ -718,12 +720,11 @@ class PagedHSTUInferLayer(torch.nn.Module):
             num_tokens: int,
             layer_input: torch.Tensor,
     ) -> torch.Tensor:
-        if not self.use_moe:
+        if not self.use_moe or self.layer_idx <= 1:
             return
         layer_input = layer_input[:num_tokens, ...]
         layer_out = self.output_buffer_[:num_tokens, ...]
 
-
         layer_out = self.sparse_moe_linear.forward(layer_out) if self.use_moe_op else self.sparse_moe_linear.forward_naive(layer_out)
 
 
diff --git a/examples/hstu/run_random_2k.sh b/examples/hstu/run_random_2k_05B.sh
similarity index 58%
rename from examples/hstu/run_random_2k.sh
rename to examples/hstu/run_random_2k_05B.sh
index 8c1e258..be5a8f7 100644
--- a/examples/hstu/run_random_2k.sh
+++ b/examples/hstu/run_random_2k_05B.sh
@@ -2,27 +2,27 @@
 export USE_COMPILE=0
 
 #开启cudagraph
-export USR_CUDA_GRAPH=1
+export USR_CUDA_GRAPH=0
 
 # 运行轮数(运行多少轮采集性能数据)
-export RUN_NUM=10
+export RUN_NUM=2000
 
 # 模型参数
-export MODEL_PATH="hstu3B.bin"
+export MODEL_PATH="hstu_05B.pth"
 export USE_MOE=0     #  USE_FFN 和 USE_MOE 不能同时开启
-export LAYER_NUM=7
-export MOE_EXPERT_TOTAL_NUM=4
-export MOE_EXPERT_ACT_NUM=2
-export USE_FFN=1     #  USE_FFN 和 USE_MOE 不能同时开启
+export LAYER_NUM=15
+export MOE_EXPERT_TOTAL_NUM=16
+export MOE_EXPERT_ACT_NUM=16
+export USE_FFN=0     #  USE_FFN 和 USE_MOE 不能同时开启
 export USE_FFN_NUMBER=2
 # 不比较精度，可以使用随机初始化参数的模型
 export USE_RANDOM_PARM=1
-export PARM_MAX_VALUE="0.5"
+export PARM_MAX_VALUE="0.1"
 
 export SEQ_LEN=2048
 export CANDIDATE_LEN=100
-export HIDDEN_DIM=4096
-export HEAD_NUM=16
+export HIDDEN_DIM=768
+export HEAD_NUM=3
 export HEAD_DIM=256
 
 export ENABLE_MATMUL_FP8=0
@@ -33,7 +33,7 @@ export USE_HSTU_OP=1
 export USE_RAB=0
 
 # prof路径
-export PROF_SAVE_PATH="prof_result_2k"
+export PROF_SAVE_PATH="prof_result_2k_05B"
 
 #batchsize 目前L20只支持1 2 4, 超过会超显存
 export BATCH_SIZE=1
@@ -48,4 +48,5 @@ export USE_MOE_OP=1
 export TEST_DYNAMIC=0
 
 #执行
-python3  benchmark/inference_benchmark_random_2k.py
\ No newline at end of file
+LOG_SUFFIX="$(date '+%Y%m%d_%H%M%S')_2k_compile${USE_COMPILE}_aclgraph${USR_CUDA_GRAPH}_precise${BATCH_SIZE}_e2e"
+python3  benchmark/inference_benchmark_random_2k.py 2>&1 | tee temp_${LOG_SUFFIX}_profiling.log
\ No newline at end of file
diff --git a/examples/hstu/run_random_2k_1B.sh b/examples/hstu/run_random_2k_1B.sh
new file mode 100644
index 0000000..455b222
--- /dev/null
+++ b/examples/hstu/run_random_2k_1B.sh
@@ -0,0 +1,53 @@
+#开启编译
+export USE_COMPILE=0
+
+#开启cudagraph
+export USR_CUDA_GRAPH=0
+
+# 运行轮数(运行多少轮采集性能数据)
+export RUN_NUM=2000
+
+# 模型参数
+export MODEL_PATH="hstu_1B.pth"
+export USE_MOE=1     #  USE_FFN 和 USE_MOE 不能同时开启
+export LAYER_NUM=15
+export MOE_EXPERT_TOTAL_NUM=16
+export MOE_EXPERT_ACT_NUM=16
+export MOE_HIDDEN_SIZE=2048
+export USE_FFN=0     #  USE_FFN 和 USE_MOE 不能同时开启
+export USE_FFN_NUMBER=2
+# 不比较精度，可以使用随机初始化参数的模型
+export USE_RANDOM_PARM=1
+export PARM_MAX_VALUE="0.1"
+
+export SEQ_LEN=2048
+export CANDIDATE_LEN=100
+export HIDDEN_DIM=768
+export HEAD_NUM=3
+export HEAD_DIM=256
+
+export ENABLE_MATMUL_FP8=0
+
+export USE_FLOAT16=0
+export USE_FLOAT32=0
+export USE_HSTU_OP=1
+export USE_RAB=0
+
+# prof路径
+export PROF_SAVE_PATH="prof_result_2k_05B"
+
+#batchsize 目前L20只支持1 2 4, 超过会超显存
+export BATCH_SIZE=1
+
+#是否加载NPU输入,维持和NPU输入一致
+export USE_NPU_INPUT=0
+
+#是否保存输入和输出tensor, 和NPU对比精度
+export SAVE_INPUT_AND_PUTPUT=1
+
+export USE_MOE_OP=1
+export TEST_DYNAMIC=0
+
+#执行
+LOG_SUFFIX="$(date '+%Y%m%d_%H%M%S')_2k_compile${USE_COMPILE}_aclgraph${USR_CUDA_GRAPH}_precise${BATCH_SIZE}_e2e"
+python3  benchmark/inference_benchmark_random_2k.py 2>&1 | tee temp_${LOG_SUFFIX}_profiling.log
\ No newline at end of file
diff --git a/examples/hstu/run_random_2k_2B.sh b/examples/hstu/run_random_2k_2B.sh
new file mode 100644
index 0000000..80ea773
--- /dev/null
+++ b/examples/hstu/run_random_2k_2B.sh
@@ -0,0 +1,53 @@
+#开启编译
+export USE_COMPILE=0
+
+#开启cudagraph
+export USR_CUDA_GRAPH=0
+
+# 运行轮数(运行多少轮采集性能数据)
+export RUN_NUM=2000
+
+# 模型参数
+export MODEL_PATH="hstu_2B.pth"
+export USE_MOE=1     #  USE_FFN 和 USE_MOE 不能同时开启
+export LAYER_NUM=15
+export MOE_EXPERT_TOTAL_NUM=16
+export MOE_EXPERT_ACT_NUM=16
+export MOE_HIDDEN_SIZE=4096
+export USE_FFN=0     #  USE_FFN 和 USE_MOE 不能同时开启
+export USE_FFN_NUMBER=2
+# 不比较精度，可以使用随机初始化参数的模型
+export USE_RANDOM_PARM=1
+export PARM_MAX_VALUE="0.1"
+
+export SEQ_LEN=2048
+export CANDIDATE_LEN=100
+export HIDDEN_DIM=768
+export HEAD_NUM=3
+export HEAD_DIM=256
+
+export ENABLE_MATMUL_FP8=0
+
+export USE_FLOAT16=0
+export USE_FLOAT32=0
+export USE_HSTU_OP=1
+export USE_RAB=0
+
+# prof路径
+export PROF_SAVE_PATH="prof_result_2k_05B"
+
+#batchsize 目前L20只支持1 2 4, 超过会超显存
+export BATCH_SIZE=1
+
+#是否加载NPU输入,维持和NPU输入一致
+export USE_NPU_INPUT=0
+
+#是否保存输入和输出tensor, 和NPU对比精度
+export SAVE_INPUT_AND_PUTPUT=1
+
+export USE_MOE_OP=1
+export TEST_DYNAMIC=0
+
+#执行
+LOG_SUFFIX="$(date '+%Y%m%d_%H%M%S')_2k_compile${USE_COMPILE}_aclgraph${USR_CUDA_GRAPH}_precise${BATCH_SIZE}_e2e"
+python3  benchmark/inference_benchmark_random_2k.py 2>&1 | tee temp_${LOG_SUFFIX}_profiling.log
\ No newline at end of file
