diff --git a/examples/hstu/modules/paged_hstu_infer_layer.py b/examples/hstu/modules/paged_hstu_infer_layer.py
index 7c283e7..4590889 100644
--- a/examples/hstu/modules/paged_hstu_infer_layer.py
+++ b/examples/hstu/modules/paged_hstu_infer_layer.py
@@ -566,7 +566,7 @@ class PagedHSTUInferLayer(torch.nn.Module):
     ):
         pass
 
-    @torch.inference_mode()
+    @torch.no_grad()
     def forward_naive(
             self,
             batch_size: int,
@@ -671,7 +671,7 @@ class PagedHSTUInferLayer(torch.nn.Module):
 
         return layer_output
 
-    @torch.inference_mode()
+    @torch.no_grad()
     def forward_input(
             self,
             batch_size: int,
@@ -702,7 +702,7 @@ class PagedHSTUInferLayer(torch.nn.Module):
 
         return self.uvqk_buffer_[:num_tokens, ...]
 
-    @torch.inference_mode()
+    @torch.no_grad()
     def forward_output(
             self,
             batch_size: int,
@@ -794,7 +794,7 @@ class PagedHSTUInferLayer(torch.nn.Module):
         self.output_buffer_[:num_tokens, ...] = layer_output
         return layer_output
 
-    @torch.inference_mode()
+    @torch.no_grad()
     def forward_moe(
             self,
             num_tokens: int,
